### ISSUE TITLE
Make sure correct python installation is used

### ISSUE THREAD

**Issue by osmatrixcode** [BienfaitKI/polus-dev#10]
I can see that there is a problem with using this package with Python installations >=3.11.7. I had an error installing the packages in requirements.txt when using a Python version that is greater than 3.11.7.  It would be nice to make a reminder to use a valid version of python for this project. Perhaps we can add a warning to this in the README file or we can write a few lines of code to raise an exception? I would love to work on this issue!

I can see in the setup.py that it requires a python version >=3.7.3, <=3.11.7

**Comment by osmatrixcode** [comment]
**Here is what I did for this problem to occur.**
- Create a conda enviroment (or just any general environment)
- Install Pip into the environment (will by default install the latest version of python)
- use Pip to install packages mentioned in the requirements.txt
- Error due to having an incorrect python version

**Comment by suzanneEmbury** [comment]
Thanks for working out the required Python version @osmatrixcode .  Here's how I got the sample code to run  on macOS Sonoma 14.6 with Python 3.11.7, in case it helps you to write a PR for this issue.  (The code is throwing some errors, and I did not let it run to completion but I don't think any of the errors are Python version problems, so I'll leave them for another issue.)

First, I installed [pyenv](https://github.com/pyenv) following the instructions closely in the [README](https://github.com/pyenv/pyenv/blob/master/README.md#unixmacos).

Pyenv is a shell tool that allows you to have multiple Python versions installed, and to specify that certain Python versions be used in certain cases.  For example, it means that I can use my normal recent Python installation in most places on my system, but when I am working in the `polus-dev` folder, when I issue the `python` command it will pick up the version I set to use in that folder (using `pyenv local 3.11.7`).

Pyenv only works for MacOS and Linux.  It can used on WSL on Windows, but apparently won't use the native Windows Python for that, so isn't going to give a true test of running the code on Windows.  There is a Windows specific fork of Pyenv ([pyenv-win](https://github.com/pyenv-win/pyenv-win)) but I haven't tried it myself yet.

On MacOS, I used Homebrew to install `pyenv` and then I followed all the shell set up commands for both bash and zsh, since I sometimes use them both.  This turned out to be necessary, as my shell couldn't pick up the correct version of Python in the `polus-dev` folder until I had run `eval "$(pyenv init -)"` (or got it to be run when the shell started up through the shell config files).

Of course, I also needed to install the required version of Python (v 3.11.7).  I did this through pyenv:

    pyenv install 3.11.7

And then I changed into my `polus-dev` source folder and told `pyenv` to use the required Python version:

    % cd ~/git/polus-dev
    % pyenv local 3.11.7
    % python --version 
    Python 3.11.7

After that, I followed the instructions in the Polus [README file](https://github.com/BienfaitKI/polus-dev/blob/master/README.md).

Then I could go into the `examples` folder and run the `script.py` file there.

HTH

**Comment by JamesDownes5** [comment]
I can run Polus as a local module with Python 3.12 but not with Python 3.13 as the packages required haven't been updated to support Python 3.13

**Comment by osmatrixcode** [comment]
> I can run Polus as a local module with Python 3.12 but not with Python 3.13 as the packages required haven't been updated to support Python 3.13

Hi yes, that is the problem I faced, with a more newer version of Python, I could not install the necessary requirements.

**Comment by osmatrixcode** [comment]
> Thanks for working out the required Python version @osmatrixcode . Here's how I got the sample code to run on macOS Sonoma 14.6 with Python 3.11.7, in case it helps you to write a PR for this issue. (The code is throwing some errors, and I did not let it run to completion but I don't think any of the errors are Python version problems, so I'll leave them for another issue.)
> 
> First, I installed [pyenv](https://github.com/pyenv) following the instructions closely in the [README](https://github.com/pyenv/pyenv/blob/master/README.md#unixmacos).
> 
> Pyenv is a shell tool that allows you to have multiple Python versions installed, and to specify that certain Python versions be used in certain cases. For example, it means that I can use my normal recent Python installation in most places on my system, but when I am working in the `polus-dev` folder, when I issue the `python` command it will pick up the version I set to use in that folder (using `pyenv local 3.11.7`).
> 
> Pyenv only works for MacOS and Linux. It can used on WSL on Windows, but apparently won't use the native Windows Python for that, so isn't going to give a true test of running the code on Windows. There is a Windows specific fork of Pyenv ([pyenv-win](https://github.com/pyenv-win/pyenv-win)) but I haven't tried it myself yet.
> 
> On MacOS, I used Homebrew to install `pyenv` and then I followed all the shell set up commands for both bash and zsh, since I sometimes use them both. This turned out to be necessary, as my shell couldn't pick up the correct version of Python in the `polus-dev` folder until I had run `eval "$(pyenv init -)"` (or got it to be run when the shell started up through the shell config files).
> 
> Of course, I also needed to install the required version of Python (v 3.11.7). I did this through pyenv:
> 
> ```
> pyenv install 3.11.7
> ```
> 
> And then I changed into my `polus-dev` source folder and told `pyenv` to use the required Python version:
> 
> ```
> % cd ~/git/polus-dev
> % pyenv local 3.11.7
> % python --version 
> Python 3.11.7
> ```
> 
> After that, I followed the instructions in the Polus [README file](https://github.com/BienfaitKI/polus-dev/blob/master/README.md).
> 
> Then I could go into the `examples` folder and run the `script.py` file there.
> 
> HTH

thank you for your reply!
It is nice to see that this isn't just caused by Conda, but with other virtual environment setups (such as pyenv)

**Comment by osmatrixcode** [comment]
My plan for this Issue is to just add a section at the top of the readme file, reminding people that you need to install a relevant version of Python for polus-dev to work. I guess then we can talk with the owner of this repository on how we can deal with this issue, and if they want to upgrade python versions + the python packages versions they need to install from the requirements.txt

**Comment by osmatrixcode** [comment]
> My plan for this Issue is to just add a section at the top of the readme file, reminding people that you need to install a relevant version of Python for polus-dev to work. I guess then we can talk with the owner of this repository on how we can deal with this issue, and if they want to upgrade python versions + the python packages versions they need to install from the requirements.txt

My main concern is that if we update the python versions and the python packages versions, it might cause an error with the results from running this codebase (might not give the correct results as expected)

**Comment by suzanneEmbury** [comment]
> My main concern is that if we update the python versions and the python packages versions, it might cause an error with the results from running this codebase (might not give the correct results as expected)

Good point.  Do we know exactly which packages are causing the problem?  I'm not sure I really understand what the root of the problem is.
